ui/app/hooks/useCurrencyDisplay.js

```diff
@@ -0,0 +1,85 @@
+
+/**
+ * Defines the shape for the Token input parameter for UseCurrencyOptions
+ * @typedef {Object} Token
+ * @property {string} symbol   - The string to use as a suffix for the token (eg. DAI)
+ * @property {decimals} number - The number of decimals to show when displaying this type of token
+ */
+
+/**
+ * Defines the shape of the options parameter for useCurrencyDisplay
+ * @typedef {Object} UseCurrencyOptions
+ * @property {token}  [Token] - When present changes the behavior to format the currency as the provided token
+ * @property {string} [displayValue] - When present is used in lieu of formatting the inputValue
+ * @property {string} [prefix] - String to prepend to the final result
+ * @property {number} [numberOfDecimals] - Number of significant decimals to display
+ * @property {string} [denomination] - Denomination (wei, gwei) to convert to for display
+ * @property {string} [currency] - Currency type to convert to. This is an override of the user's nativeCurrency
+ */
+
+/**
+ * Defines the return shape of the second value in the tuple
+ * @typedef {Object} CurrencyDisplayParts
+ * @property {string} [prefix]  - string to prepend to the value for display
+ * @property {string} value     - string representing the value, formatted for display
+ * @property {string} [suffix]  - string to append to the value for display
+ */
+
+/**
+ * useCurrencyDisplay hook
+ *
+ * Given a hexadecimal encoded value string OR the decrypted data object from a token transaction,
+ * and an object of parameters used for formatting the display, produce both a fully formed string
+ * and the pieces of that string used for displaying the currency to the user
+ * @param {string | Object} inputValue - The value to format for display
+ * @param {UseCurrencyOptions} opts    - An object for options to format the inputValue
+ * @return {[string, CurrencyDisplayParts]}
+ */
+export function useCurrencyDisplay (inputValue, { token, displayValue, prefix, numberOfDecimals, denomination, currency, ...opts }) {
+  const { currentCurrency, nativeCurrency, conversionRate } = useSelector(
+    ({ metamask: { currentCurrency, nativeCurrency, conversionRate } }) => ({ currentCurrency, nativeCurrency, conversionRate })
+  )
+
+  const toCurrency = currency || currentCurrency
+
+  const value = useMemo(() => {
+    if (typeof inputValue === 'object') {
+      if (token === null) {
+        throw new Error(
+          `useCurrencyDisplay provided an inputValue that is an object, without a corresponding opts.token.
+          To format a token for display provide the token object in the options object`
+        )
+      }
+      if ((inputValue?.params?.length || 0) === 0) {
+        throw new Error(
+          `useCurrencyDisplay provided an object without params. Only use this hook with decoded tokenData, or a value string`
+        )
+      }
+      const tokenValue = getTokenValue(inputValue.params)
+      return calcTokenAmount(tokenValue, token.decimals).toString()
```

Comment:
Where did this token-related logic come from? I can't see anything like this in the container being replaced.

Reply:
Good callout. The token related logic was added as part of my work on adding the new transaction list components. It allows for the token used in the transaction to be displayed in the history item (without seletectedToken from the redux state) eg: `Send DAI    - 5 DAI`

If it would make for a more concise git history I could extract the bits that do not match directly to logic from the old container and move it into the other PR?

Reply:
Certainly keeping this PR as a pure refactor would make it a lot easier to review, so that would be appreciated.

Though this logic does look a lot like `tokenCurrencyDisplay` Perhaps it should be made into a separate hook, instead of putting it all into the `useCurrencyDisplay` hook? Or maybe we could have a separate hook with the two combined, for cases where it's easier to render either depending on what's given? It does seem a bit strange to introduce token logic in every place `CurrencyDisplay` is currently used, where in many cases we know it's not a token amount being displayed.

Reply:
@Gudahtt you're absolutely right. Here's the reasoning behind why it was included here:

originally the transaction-list-item component renders either the CurrencyDisplay or TokenCurrencyDisplay component depending on whether or not the user has selected a token. The new designs actually change this behavior to be based on the transaction type rather than what token the user has selected. To my knowledge and grep the TokenCurrencyDisplay component is only ever used here.

The main driver for having the token in the useCurrencyDisplay hook was due to not being able to conditionally call hooks. I was still unsure of the overall interlinking of CurrencyDisplay and UserPreferencedCurrencyDisplay and TokenCurrencyDisplay. 

--- 

What I have done to address your feedback is to remove the token display functionality into a separate hook, and then update the TokenCurrencyDisplay component to use it, very similar to the UserPreferencedCurrencyDisplay component. I intend on removing the TokenCurrencyDisplay component in my next PR because it will no longer be in use. 

Added tests and documentation for the new useTokenDisplayValue hook, which is intended to either be passed to CurrencyDisplay component as it is in TokenCurrencyDisplay component or passed to useCurrencyDisplay hook in a similar fashion.

Reply:
Sounds great! I'll take a look soon.